File: src/com/demo/leetcode/onethousand/Leetcode234.java
package com.demo.leetcode.onethousand;

import com.demo.leetcode.datastruct.ListNode;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.LinkedList;

public class Leetcode234 {

    public static void main(String[] args) {
        ListNode node = new ListNode(1);
        node.next = new ListNode(2);
        node.next.next = new ListNode(2);
        node.next.next.next = new ListNode(1, null);
        System.out.println(isPalindrome(node));
    }

    public static boolean isPalindrome(ListNode head) {
        Deque stack = new LinkedList();
        ListNode temp = head;
        stack.push(temp);
        while (temp.next != null) {
            temp = temp.next;
            stack.push(temp);
        }
        ListNode n = null;
        ListNode cur = null;
        while (stack.size() != 0) {
            ListNode node = (ListNode) stack.poll();
            if (n == null) {
                n = new ListNode(node.val);
                cur = n;
            } else {
                cur.next = new ListNode(node.val);;
                cur = cur.next;
            }
        }
        while (head.next != null) {
            if (n.val != head.val) {
                return false;
            } else {
                head = head.next;
                n = n.next;
            }

        }
        return true;
    }
}

File: src/com/demo/leetcode/onethousand/Leetcode234.kt
package com.demo.leetcode.onethousand

import com.demo.leetcode.datastruct.ListNode
import java.util.*

fun main() {
    val node = ListNode(1)
    node.next = ListNode(2)
    node.next.next = ListNode(2)
    node.next.next.next = ListNode(1, null)
    println(isPalindrome(node))
}

fun isPalindrome(head: ListNode?): Boolean {
    val stack = LinkedList<ListNode?>()
    var temp = head
    stack.push(temp)
    while (temp?.next != null) {
        temp = temp.next
        stack.push(temp)
    }
    var n: ListNode? = null
    var cur: ListNode? = null
    while (stack.size != 0) {
        val node = stack.poll() as ListNode
        if (n == null) {
            n = ListNode(node.`val`)
            cur = n
        } else {
            cur!!.next = ListNode(node.`val`)
            cur = cur!!.next
        }
    }
    var h = head
    while (h?.next != null) {
        if (n!!.`val` != h.`val`) {
            return false
        } else {
            h = h.next
            n = n.next
        }
    }
    return true
}